src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/SystemResources.cs

```diff
@@ -2028,6 +2029,123 @@ internal override bool IsUnset
         #endregion Properties
     }
 
+    internal class DeferredResourceReferenceList : IEnumerable
+    {
+        private readonly object _syncRoot = new();
+        private readonly Dictionary<object, WeakReference<DeferredResourceReference>> _entries = new();
+        private int _potentiallyDeadEntryCount = 0;
+
+        public void AddOrSet(DeferredResourceReference deferredResourceReference)
+        {
+            lock (_syncRoot)
+            {
+                _entries[deferredResourceReference.Key] = new WeakReference<DeferredResourceReference>(deferredResourceReference);
+            }
+        }
+
+        public void Remove(DeferredResourceReference deferredResourceReference)
+        {
+            lock (_syncRoot)
+            {
+                _entries.Remove(deferredResourceReference.Key);
+            }
+        }
+
+        internal DeferredResourceReference Get(object resourceKey)
+        {
+            WeakReference<DeferredResourceReference> weakReference;
+            lock (_syncRoot)
+            {
+                _entries.TryGetValue(resourceKey, out weakReference);
+            }
+
+            if (weakReference is null)
+            {
+                return null;
+            }
+
+            if (weakReference.TryGetTarget(out DeferredResourceReference deferredResourceReference))
+            {
+                return deferredResourceReference;
+            }
+            else
+            {
+                ++_potentiallyDeadEntryCount;
```

Comment:
Should this be `Interlocked.Increment`? The use of `_syncRoot` to protect all accesses to `_entries` implies that this class is used in a multithreaded context, and a simple increment is not threadsafe.

Reply:
Even with a `lock`, there's a race condition with `Purge` in this method: if `Purge` runs (on another thread) while this thread is between lines 2059 and 2073, the entry will be removed from `_entries` but then this thread will increment `_potentiallyDeadEntryCount` even though the entry no longer exists.